spring-integration-file/src/main/java/org/springframework/integration/file/splitter/FileSplitter.java

```diff
@@ -224,6 +226,23 @@ public Object next() {
 		}
 	}
 
+	@Override
+	protected void addHeaders(Message<?> message, AbstractIntegrationMessageBuilder<?> builder) {
+		if (message.getPayload() instanceof File) {
+			File file = (File) message.getPayload();
+			builder
+				.setHeaderIfAbsent(FileHeaders.ORIGINAL_FILE, file)
+				.setHeaderIfAbsent(FileHeaders.FILENAME, file.getName());
+
+		}
+		else if (message.getPayload() instanceof String) {
+			File file = new File((String) message.getPayload());
```

Comment:
Well, I see your point now Gary: we have `File` on the input, but don't have on the output, so we really may miss `fileName` for the downstream logic, if we haven't care about headers manually.

But how about to do that via some different way (`AbstractMessageSplitter`):
````java
private MessageProcessor<Map<String, Object>> headersCallback;
````
and create `headersToAdd` only once before iteration.
WDYT?